pypesto/optimize/optimizer.py

```diff
@@ -722,6 +727,94 @@ def is_least_squares(self):
         return False
 
 
+class PyswarmsOptimizer(Optimizer):
+    """
+    Global optimization using pyswarms.
+    Package homepage: https://pyswarms.readthedocs.io/en/latest/index.html
+
+    Parameters
+    ----------
+    par_popsize:
+        number of particles in the swarm, default value 10
+
+    options:
+        Optimizer options that are directly passed on to pyswarms.
+        c1: cognitive parameter
+        c2: social parameter
+        w: inertia parameter
+        Default values are (c1,c2,w) = (0.5, 0.3, 0.9)
+
+    Examples
+    --------
+    Arguments that can be passed to options:
+
+    maxiter:
+        used to calculate the maximal number of funcion evaluations.
+        Default: 1000
+    """
+
+    def __init__(self, par_popsize: float = 10, options: Dict = None):
+        super().__init__()
+
+        if options is None:
+            options = {'maxiter': 1000, 'c1': 0.5, 'c2': 0.3, 'w': 0.9}
+        self.options = options
+        self.par_popsize = par_popsize
+
+    @fix_decorator
+    @time_decorator
+    @history_decorator
+    def minimize(
+            self,
+            problem: Problem,
+            x0: np.ndarray,
+            id: str,
+            history_options: HistoryOptions = None,
+    ) -> OptimizerResult:
+        lb = problem.lb
+        ub = problem.ub
+
+        if pyswarms is None:
+            raise ImportError(
+                "This optimizer requires an installation of pyswarms.")
+
+        optimizer = pyswarms.single.global_best.GlobalBestPSO(
+            n_particles=self.par_popsize, dimensions=len(x0),
+            options=self.options, bounds=(lb, ub))
+
+        def successively_working_fval(swarm: np.ndarray) -> np.ndarray:
+            """Evaluate the function for all parameters in the swarm object.
+
+            Parameters:
+            -----------
+            swarm: np.ndarray, shape (n_particales_in_swarm, n_parameters)
+
+            Returns:
+            --------
+            result: np.ndarray, shape (n_particles_in_swarm)
+            """
+
+            n_particles = swarm.shape[0]
+            result = np.zeros(n_particles)
+            for iPar, par in enumerate(swarm):
+                result[iPar] = problem.objective.get_fval(par)
```

Comment:
This doesn't make sense to me:
In line 798, result is an array which is as long as we have particles. In line 800, it's assumed to have the size of the parameter vector... This looks like a bug.

Reply:
No, it should be correct, since ```enumerate(swarm)``` has the dimensions ```number_particles_in_swarm x number_parameters```.
In the for-loop, we assign ```iPar = number_particles_in_swarm``` and ```par = parameter```.
Since ```result = np.zeros(n_particles)```, we can index it by all number of particles and assign the value ```probem.objective.get_fval(par)```

Reply:
Ah, thanks!
So, for my understanding: "swarm" is an N times M matrix. What does enumerate then do? Does it iterate over rows, columns, or elements? I would guess it iterates over columns, correct? Please just add a line of comment, that you actually iterate over columns, like "# iterate over found parameter vector, i.e., the columns in swarm"

For me, iPar always iterates over entries in the parameter vector, this is probably why I got confused. If iPar for you iterates over "individuals in the population/swarm", maybe go for iParVector, or iVector, or iSolution... If that's done, I'm fine!

Reply:
enumerate wraps around a `for x in ...`, adding counters. when iterating over numpy arrays, iteration is always over the first dimension (=rows), i.e. over the particles, returning a parameter vector as `par`. Looks correct to me. Paul is right, iPar should be changed to `i_particle` or sth like that (lower case all).